### PR TITLE
Bi-directional reactions to messages

### PR DESCRIPTION
**Pull request by kable-wilmoth**
@RocketChat/core 

Added the ability to react to slack and rocket messages.  Doesn't quite work all the way because Slack sees the reactions as coming from the 'bot'.  I am working w/ the Slack team to figure out some work arounds based on some alternative bot authentication mechanisms.

* AddReact
    * slack message in slack: OK
    * rocket message in rocket:  BOT in slack
    * slack message in rocket: BOT in slack
    * rocket message in slack: OK
* RemoveReact
    * slack message, reacted in slack, removed in slack: OK
    * slack message, reacted in slack, removed in rocket: Removed in Rocket, BOT still in Slack
    * rocket message, reacted in rocket, removed in rocket: OK
    * rocket message, reacted in rocket, removed in slack: OK

### CODE REVIEW

File: packages/rocketchat-lib/server/models/Messages.coffee [hunk 1]
```diff
@@ -210,6 +210,12 @@ RocketChat.models.Messages = new class extends RocketChat.models._Base
 
 		return @findOne query
 
+	findOneBySlackTs: (slackTs) ->
```
Comment (kable-wilmoth) [inline packages/rocketchat-lib/server/models/Messages.coffee]: Needed the ability to look up a slack message w/ just the timestamp because the react message coming from slack didn't include the bot ID

File: packages/rocketchat-reactions/setReaction.js [hunk 2]
```diff
@@ -1,7 +1,12 @@
 /* globals msgStream */
 Meteor.methods({
-	setReaction(reaction, messageId) {
-		if (!Meteor.userId()) {
+	setReaction(reaction, messageId, user) {
```
Comment (kable-wilmoth) [inline packages/rocketchat-reactions/setReaction.js]: It was previously assumed that a reaction was coming from the current user.  This is true most of the time, but in my case when Slack sends a reaction, I needed to call this method on behalf of the user.  So added a user parameter, but if not present, it behaves as before.

File: packages/rocketchat-reactions/setReaction.js [hunk 3]
```diff
@@ -41,8 +44,10 @@ Meteor.methods({
 			if (_.isEmpty(message.reactions)) {
 				delete message.reactions;
 				RocketChat.models.Messages.unsetReactions(messageId);
```
Comment (kable-wilmoth) [inline packages/rocketchat-reactions/setReaction.js]: Added two new callbacks: setReaction and unsetReaction

File: packages/rocketchat-reactions/setReaction.js (hunk 2 above)
Comment (graywolf336) [inline packages/rocketchat-reactions/setReaction.js]: Please revert this change and instead run this method as the user who is doing the reacting like such:

```javascript
Meteor.runAsUser(user._id, () => {
	Meteor.call('setReaction', reaction, messageId);
});
```

Otherwise this is a breaking change for all of the clients out there who call this method as the user and don't pass in the user object.

File: packages/rocketchat-reactions/setReaction.js (hunk 3 above)
Comment (graywolf336) [inline packages/rocketchat-reactions/setReaction.js]: Can you add the callbacks to run on the client side as well? This way people can take advantage of it for statistics, such as Piwiki or future Google Analytics

File: packages/rocketchat-reactions/setReaction.js (hunk 2 above)
Comment (kable-wilmoth) [inline packages/rocketchat-reactions/setReaction.js]: Sure will.

Comment (graywolf336) [inline packages/rocketchat-reactions/setReaction.js]: Thanks!

File: packages/rocketchat-reactions/setReaction.js (hunk 3 above)
Comment (kable-wilmoth) [inline packages/rocketchat-reactions/setReaction.js]: Haven't done any client side changes to Rocket yet, but will look into this.

**Comment by kable-wilmoth** [comment]
@graywolf336 I have submitted the requested fixes.  Not sure how to trigger the next 'steps' in the pull request process.  Maybe this comment is it.
Thanks.
-kable
